File: dns_bench.py
#!/usr/bin/env python3
"""DNS benchmark.

Times A-record lookups across a list of public resolvers and renders a
live table that stays sorted by response time as each result comes in.

Requires dnspython:  pip install dnspython
"""

import socket
import sys
import time

import dns.exception
import dns.resolver

DEFAULT_HOST = "www.google.com"

SERVERS = {
    "OpenDNS_1": "208.67.222.222",
    "OpenDNS_2": "208.67.220.220",
    "L3_1": "209.244.0.3",
    "L3_2": "209.244.0.4",
    "Verisign": "64.6.64.6",
    "Google_1": "8.8.8.8",
    "Google_2": "8.8.4.4",
    "Quad9_1": "9.9.9.9",
    "CloudFlare": "1.1.1.1",
    "Quad9_2": "149.112.112.112",
    "DNSINC_1": "216.146.35.35",
    "DNSINC_2": "216.146.36.36",
    "CensurFriDNS": "89.233.43.71",
    "DNSWatch_1": "84.200.69.80",
    "DNSWatch_2": "84.200.70.40",
    "Hurricane Electric": "74.82.42.42",
    "OpenNIC": "94.247.43.254",
    "DNS4EU Protective": "86.54.11.1",
    "DNS4EU Child": "86.54.11.12",
    "DNS4EU Adblock": "86.54.11.13",
    "DNS4EU Unfiltered": "86.54.11.100",
}


def parse_args(argv):
    """Return the hostname to look up, or print help and exit."""
    if argv and argv[0] == "--help":
        print(f"\nUsage: {sys.argv[0]} [lookup_hostname]")
        print("If no hostname is provided www.google.com will be used\n")
        sys.exit(0)
    return argv[0] if argv else DEFAULT_HOST


def query_server(host, ip, timeout=2):
    """Time a single A-record lookup against one nameserver.

    Returns elapsed seconds, or raises a DNS/OS exception on failure.
    """
    resolver = dns.resolver.Resolver(configure=False)
    resolver.nameservers = [ip]
    resolver.timeout = timeout      # per-request timeout
    resolver.lifetime = timeout     # total time budget (covers the retry)

    start = time.perf_counter()
    resolver.resolve(host, "A")
    return time.perf_counter() - start


class LiveTable:
    """Reprints a time-sorted table in place each time a row is added."""

    def __init__(self, host):
        self.host = host
        self.rows = []          # list of (name, ip, elapsed)
        self._prev_lines = 0

    def add(self, name, ip, elapsed):
        self.rows.append((name, ip, elapsed))
        self._render()

    def _render(self):
        ordered = sorted(self.rows, key=lambda r: r[2])

        out = f"\nTiming lookups for {self.host}\n\n"
        out += f"{'Server':<20} {'IP':<15} {'Time':>4}\n"
        out += "-" * 45 + "\n"
        for name, ip, elapsed in ordered:
            out += f"{name:<20} {ip:<15} {elapsed:.5f}\n"

        # Move the cursor up over the previous render and clear it so the
        # table refreshes in place instead of scrolling.
        if self._prev_lines:
            sys.stdout.write(f"\033[{self._prev_lines}A\033[J")
        sys.stdout.write(out)
        sys.stdout.flush()
        self._prev_lines = out.count("\n")


def main():
    host = parse_args(sys.argv[1:])
    table = LiveTable(host)
    errors = []

    # OS default resolver
    try:
        start = time.perf_counter()
        socket.gethostbyname(host)
        table.add("OS_Default", "local", time.perf_counter() - start)
    except OSError as exc:
        errors.append(f"{'OS_Default':<10} {'local':<15} failed: {exc}")

    for name, ip in SERVERS.items():
        try:
            elapsed = query_server(host, ip)
        except (dns.exception.DNSException, OSError) as exc:
            msg = str(exc) or exc.__class__.__name__
            errors.append(f"{name:<10} {ip:<15} failed: {msg}")
            continue
        table.add(name, ip, elapsed)

    print()
    for err in errors:
        print(err)
    print()


if __name__ == "__main__":
    main()

File: dns_bench.pl
#!/usr/bin/perl

use strict;
use warnings;
use Net::DNS;
use Time::HiRes;
use Socket;

my $number_args = $#ARGV + 1;
my $host = "";

my $arg = $ARGV[0];
chomp($arg) if defined $arg;

if ($number_args == 1)
{
	if ($arg eq '--help')
	{
		print "\nUsage: ".$0. " [lookup_hostname]\n";
		print "If no hostname is provided www.google.com will be used\n\n";
		exit;
	}
	else
	{
		$host = $ARGV[0];
	}
}
else
{
	$host = 'www.google.com';
}

my $errorList = "";

my %servers = (
	'OpenDNS_1' => '208.67.222.222',
	'OpenDNS_2' => '208.67.220.220',
	'DNSPrivacy' => '94.130.110.185',
	'L3?' => '209.244.0.3',
	'L3?' => '209.244.0.4',
	'Verisign' => '64.6.64.6',
	'Google_1' => '8.8.8.8',
	'Google_2' => '8.8.4.4',
	'Quad9_1' => '9.9.9.9',
	'CloudFlare' => '1.1.1.1',
	'Quad9_2' => '149.112.112.112',
	'DNS.com?' => '8.26.56.26', 
	'DNS.com?' => '8.20.247.20',
	'DNSINC_1' => '216.146.35.35', 
	'DNSINC_2' => '216.146.36.36', 
	'EMERION?' => '37.235.1.174', 
	'EMERION?' => '37.235.1.177', 
	'CensurFriDNS' => '89.233.43.71', 
	'DNSWatch_1' => '84.200.69.80', 
	'DNSWatch_2' => '84.200.70.40', 
	'Hurricane Electric' => '74.82.42.42', 
	'guifi-net?' => '109.69.8.51',
	'OpenNIC' => '94.247.43.254',
    'DNS4EU Protective' => '86.54.11.1',
    'DNS4EU Child' => '86.54.11.12',
    'DNS4EU Adblock' => '86.54.11.13',
    'DNS4EU Unfiltered' => '86.54.11.100'
	);

my @results;        # successful lookups so far: { name, ip, time }
my $prev_lines = 0; # how many lines the last render printed

$| = 1;             # autoflush so the live updates show immediately

# Redraw the whole table, sorted by time, overwriting the previous render.
sub render
{
	my @sorted = sort { $a->{time} <=> $b->{time} } @results;

	my $out = "\nTiming lookups for $host\n\n";
	$out .= sprintf("%-20s %-15s %4s\n", "Server", "IP", "Time");
	$out .= ("-" x 45) . "\n";
	for my $r (@sorted)
	{
		$out .= sprintf("%-20s %-15s %.5f\n", $r->{name}, $r->{ip}, $r->{time});
	}

	# Move the cursor back up over the previous render and clear it,
	# so the table updates in place instead of scrolling.
	print "\033[${prev_lines}A\033[J" if $prev_lines > 0;
	print $out;
	$prev_lines = ($out =~ tr/\n//);
}

# OS default resolver
my $start = Time::HiRes::gettimeofday();
inet_ntoa(inet_aton($host));
my $end = Time::HiRes::gettimeofday();
push @results, { name => 'OS_Default', ip => 'local', time => $end - $start };
render();

while (my ($name, $ip) = each(%servers))
{
	#todo: this should be a function
	my $res = Net::DNS::Resolver->new(nameservers => [$ip]);
	$res->udp_timeout(2);
	$res->retry(1);
	my $start = Time::HiRes::gettimeofday();
	my $query = $res->search($host);
	my $end = Time::HiRes::gettimeofday();

	if ($query)
	{
		foreach my $rr ($query->answer)
		{
			next unless $rr->type eq "A";
			#print $rr->address, "\n";
		}

		push @results, { name => $name, ip => $ip, time => $end - $start };
		render();
	}
	else
	{
		my $err = sprintf("%-10s %-15s failed: %s \n", $name, $ip, $res->errorstring);
		$errorList .= $err;
	}
}

print "\n", $errorList, "\n";
